xen/arch/arm/coproc/schedule.c

```diff
@@ -0,0 +1,317 @@
+	unsigned long flags;
+
+	if ( !vcoproc_scheduler_vcoproc_is_destroyed(sched, vcoproc) )
+		return -EINVAL;
+
+	spin_lock_irqsave(&sched_data->schedule_lock, flags);
+	vcoproc->sched_priv = VCOPROC_SCHED_OP(sched, alloc_vdata, vcoproc);
+	if ( !vcoproc->sched_priv )
+	{
+		printk("Failed to allocate scheduler-specific data for vcoproc \"%s\"\n",
+				dev_path(vcoproc->coproc->dev));
+		spin_unlock_irqrestore(&sched_data->schedule_lock, flags);
+		return -ENOMEM;
+	}
+
+	vcoproc->state = VCOPROC_SLEEPING;
+	spin_unlock_irqrestore(&sched_data->schedule_lock, flags);
+
+	return 0;
+}
+
+int vcoproc_scheduler_vcoproc_destroy(struct vcoproc_scheduler *sched,
+		struct vcoproc_instance *vcoproc)
+{
+	struct vcoproc_schedule_data *sched_data = sched->sched_priv;
+	unsigned long flags;
+
+	if ( vcoproc_scheduler_vcoproc_is_destroyed(sched, vcoproc) )
+		return 0;
+
+	vcoproc_sheduler_vcoproc_sleep(sched, vcoproc);
+
+	spin_lock_irqsave(&sched_data->schedule_lock, flags);
+	if ( vcoproc->state == VCOPROC_TERMINATING )
+	{
+		spin_unlock_irqrestore(&sched_data->schedule_lock, flags);
+		return -EBUSY;
+	}
+
+	if ( vcoproc->sched_priv )
+	{
+		VCOPROC_SCHED_OP(sched, free_vdata, vcoproc->sched_priv);
+		vcoproc->sched_priv = NULL;
+	}
+	vcoproc->state = VCOPROC_UNKNOWN;
+	spin_unlock_irqrestore(&sched_data->schedule_lock, flags);
+
+	return 0;
+}
+
+void vcoproc_sheduler_vcoproc_wake(struct vcoproc_scheduler *sched, struct vcoproc_instance *vcoproc)
+{
+	struct vcoproc_schedule_data *sched_data = sched->sched_priv;
+	unsigned long flags;
+
+	ASSERT(!vcoproc_scheduler_vcoproc_is_destroyed(sched, vcoproc));
+
+	/* TODO What to do if we came with state TERMINATING? */
+	spin_lock_irqsave(&sched_data->schedule_lock, flags);
+	if ( vcoproc->state != VCOPROC_SLEEPING )
+	{
+		spin_unlock_irqrestore(&sched_data->schedule_lock, flags);
+		return;
+	}
+
+	VCOPROC_SCHED_OP(sched, wake, vcoproc);
+	vcoproc->state = VCOPROC_WAITING;
+	spin_unlock_irqrestore(&sched_data->schedule_lock, flags);
+
+	vcoproc_schedule(sched);
+}
+
+void vcoproc_sheduler_vcoproc_sleep(struct vcoproc_scheduler *sched, struct vcoproc_instance *vcoproc)
+{
+	struct vcoproc_schedule_data *sched_data = sched->sched_priv;
+	unsigned long flags;
+	int reschedule = 0;
+
+	ASSERT(!vcoproc_scheduler_vcoproc_is_destroyed(sched, vcoproc));
+
+	spin_lock_irqsave(&sched_data->schedule_lock, flags);
+	if ( vcoproc->state != VCOPROC_WAITING && vcoproc->state != VCOPROC_RUNNING )
+	{
+		spin_unlock_irqrestore(&sched_data->schedule_lock, flags);
+		return;
+	}
+
+	VCOPROC_SCHED_OP(sched, sleep, vcoproc);
+	if ( vcoproc->state == VCOPROC_WAITING )
+		vcoproc->state = VCOPROC_SLEEPING;
+	else
+	{
+		vcoproc->state = VCOPROC_TERMINATING;
+		reschedule = 1;
+	}
+	spin_unlock_irqrestore(&sched_data->schedule_lock, flags);
+
+	if ( reschedule )
+		vcoproc_schedule(sched);
+}
+
+void vcoproc_sheduler_vcoproc_yield(struct vcoproc_scheduler *sched, struct vcoproc_instance *vcoproc)
```

Comment:
From my point of view the yield function like wake/sleep doesn't need a return value, since state is being changed inside function. At least until I won't see "the whole picture".